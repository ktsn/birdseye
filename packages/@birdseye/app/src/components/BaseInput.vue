<template>
  <div class="base-input">
    <component
      v-if="!removeInput"
      :is="componentForType"
      :value="value"
      @input="$emit('input', arguments[0])"
    />

    <select
      v-if="!removeTypes"
      :value="typeOfValue"
      class="select-type"
      @change="$emit('change-type', $event.target.value)"
    >
      <option
        v-for="type in realAvailableTypes"
        :key="type"
        :value="type"
      >{{ type }}</option>
    </select>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import InputString from './InputString.vue'
import InputNumber from './InputNumber.vue'
import InputBoolean from './InputBoolean.vue'
import InputArray from './InputArray.vue'
import InputObject from './InputObject.vue'

const typeToComponentName: Record<string, string> = {
  string: 'InputString',
  number: 'InputNumber',
  boolean: 'InputBoolean',
  array: 'InputArray',
  object: 'InputObject'
}

const possibleTypes = Object.keys(typeToComponentName)

export default Vue.extend({
  name: 'BaseInput',

  components: {
    InputString,
    InputNumber,
    InputBoolean,
    InputArray,
    InputObject
  },

  props: {
    value: {
      type: [String, Number, Boolean, Array, Object],
      default: undefined
    },

    availableTypes: {
      type: Array as () => string[],
      default: () => [],
      validator(types: string[]) {
        return types.every(type => possibleTypes.indexOf(type) >= 0)
      }
    },

    removeInput: {
      type: Boolean,
      default: false
    },

    removeTypes: {
      type: Boolean,
      default: false
    }
  },

  computed: {
    /**
     * Replace empty array to possibleTypes
     */
    realAvailableTypes(): string[] {
      return this.availableTypes.length === 0
        ? possibleTypes
        : this.availableTypes
    },

    typeOfValue(): string {
      if (this.value === null) {
        return 'null'
      }

      if (this.value === undefined) {
        return 'undefined'
      }

      if (Array.isArray(this.value)) {
        return 'array'
      }

      return typeof this.value
    },

    componentForType(): string {
      return typeToComponentName[this.typeOfValue]
    }
  }
})
</script>

<birdseye lang="yml">
name: BaseInput
patterns:
  - name: String
    props:
      value: string value
  - name: Number
    props:
      value: 123
  - name: Boolean
    props:
      value: true
  - name: Array
    props:
      value:
        - foo
        - 42
        - true
      removeTypes: true
  - name: Nested Array
    props:
      value:
        - foo
        - [1, 2, 3]
        - [bar, baz]
      removeTypes: true
  - name: Object
    props:
      value:
        a: foo
        b: 42
        c: true
      removeTypes: true
  - name: Nested Object
    props:
      value:
        a: foo
        b:
          x: 1
          y: 2
          z: 3
        c:
          test1: bar
          test2: baz
      removeTypes: true
</birdseye>
