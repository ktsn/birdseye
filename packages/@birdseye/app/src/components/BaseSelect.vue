<template>
  <select
    :value="value"
    class="base-select"
    @change="$emit('change', $event.target.value)"
  >
    <slot />
  </select>
</template>

<script lang="ts">
import Vue from 'vue'

export default Vue.extend({
  name: 'BaseSelect',

  props: {
    value: {
      type: String,
      required: true
    }
  }
})
</script>

<style scoped>
.base-select {
  appearance: none;
  -moz-appearance: none;
  -webkit-appearance: none;
  background: none;
  border: none;
  color: #aaa;
  text-decoration: underline;
  cursor: pointer;
  outline: none;
  transition: color 150ms var(--ease-out-cubic);
}

.base-select:hover,
.base-select:focus {
  color: var(--color-main);
}
</style>
