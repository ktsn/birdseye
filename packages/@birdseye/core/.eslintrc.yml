---
  root: true
  extends: ktsn-vue
  rules:
    typescript/no-type-alias: off