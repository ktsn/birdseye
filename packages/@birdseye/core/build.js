const pkg = require('./package.json')
const { rollup } = require('rollup')
const vue = require('rollup-plugin-vue').default
const css = require('rollup-plugin-css-only')
const ts = require('rollup-plugin-typescript2')

const banner = `/*!
* ${pkg.name} v${pkg.version}
* ${pkg.homepage}
*
* @license
* Copyright (c) 2018 ${pkg.author}
* Released under the MIT license
*/`

function capitalize(name) {
  const camelized = name.replace(/[-_](\w)/g, (_, c) => c.toUpperCase())
  return camelized[0].toUpperCase() + camelized.slice(1)
}

const base = {
  input: 'src/index.ts',
  output: {
    banner,
    exports: 'named',
    moduleName: capitalize(pkg.name),
    global: {
      vue: 'Vue'
    }
  },

  external: ['vue'],
  plugins: [
    vue({
      css: false
    }),

    ts({
      tsconfig: './tsconfig.main.json',
      clean: true,
      typescript: require('typescript')
    }),

    css({
      output: 'dist/core.css'
    })
  ]
}

async function build(type) {
  const start = Date.now()

  const bundle = await rollup(base)

  const file = `dist/core.${type}.js`
  await bundle.write(
    Object.assign({}, base.output, {
      file,
      format: type
    })
  )

  const end = Date.now()

  console.log(`\u001b[32mcreated ${file} in ${(end - start) / 1000}s\u001b[0m`)
}

build('cjs')
  .then(() => build('es'))
  .catch(err => {
    console.error(String(err))
    process.exit(1)
  })
