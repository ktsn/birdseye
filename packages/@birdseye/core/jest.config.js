module.exports = {
  transform: {
    '^.+\\.[jt]s$': 'ts-jest',
    '^.+\\.vue$': 'vue-jest'
  },
  setupFiles: ['<rootDir>/test/setup.ts'],
  testURL: 'http://localhost',
  testRegex: '/test/.+\\.spec\\.(js|ts)$',
  moduleNameMapper: {
    '^@/(.+)$': '<rootDir>/src/$1',
    '^vue$': 'vue/dist/vue.runtime.common.js'
  },
  moduleFileExtensions: ['ts', 'js', 'json', 'vue'],
  collectCoverageFrom: ['src/**/*.{ts,vue}'],
  globals: {
    'ts-jest': {
      tsConfigFile: 'tsconfig.test.json'
    }
  }
}
