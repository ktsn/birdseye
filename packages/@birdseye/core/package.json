{
  "name": "@birdseye/core",
  "version": "0.1.0",
  "author": "katashin",
  "description": "Core modules of Birdseye",
  "keywords": [
    "Birdseye",
    "core"
  ],
  "license": "MIT",
  "main": "dist/core.cjs.js",
  "module": "dist/core.es.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "homepage": "https://github.com/ktsn/birdseye",
  "bugs": "https://github.com/ktsn/birdseye/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/ktsn/birdseye.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "prepare": "yarn clean && yarn build",
    "prepublishOnly": "yarn test",
    "clean": "rm -rf dist",
    "play": "poi --open --config play.config.js",
    "build": "node build.js",
    "build:play": "poi build --config play.config.js",
    "lint": "eslint --ext js,ts,vue src test",
    "lint:fix": "yarn lint --fix",
    "test": "yarn lint && yarn test:unit",
    "test:unit": "jest",
    "test:dev": "jest --watch"
  },
  "devDependencies": {
    "@babel/core": "^7.0.0-beta.55",
    "@types/jest": "^23.3.0",
    "@vue/test-utils": "^1.0.0-beta.20",
    "babel-core": "7.0.0-bridge.0",
    "eslint": "^5.1.0",
    "eslint-config-ktsn-vue": "^1.0.1",
    "jest": "^23.4.1",
    "poi": "^10.2.9",
    "postcss": "^7.0.2",
    "prettier": "1.14.0",
    "prettier-config-ktsn": "^1.0.0",
    "rollup": "^0.63.4",
    "rollup-plugin-css-only": "^0.4.0",
    "rollup-plugin-typescript2": "^0.16.1",
    "rollup-plugin-vue": "^4.3.1",
    "ts-jest": "^23.0.1",
    "ts-loader": "^4.4.2",
    "typescript": "^2.9.2",
    "vue": "^2.5.16",
    "vue-jest": "^2.6.0",
    "vue-play": "^3.2.1",
    "vue-template-compiler": "^2.5.16"
  }
}