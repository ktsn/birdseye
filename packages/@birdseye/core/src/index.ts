export * from './interfaces'
export { normalizeMeta } from './meta'
