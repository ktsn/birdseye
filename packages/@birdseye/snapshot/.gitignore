node_modules/
/lib/
/birdseye/snapshots