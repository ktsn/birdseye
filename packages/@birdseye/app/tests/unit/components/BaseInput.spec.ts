import { shallowMount, Wrapper } from '@vue/test-utils'
import BaseInput from '@/components/BaseInput.vue'
import BaseSelect from '@/components/BaseSelect.vue'
import { ComponentDataType } from '@birdseye/core'

describe('BaseInput', () => {
  describe('rendering', () => {
    it('removes input', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: 'foo',
          removeInput: true
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('removes type selector', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: 'foo',
          removeTypes: true
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('supported types', () => {
    it('string', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: 'string value'
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('number', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: 123
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('boolean', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: true
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('array', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: ['foo', 42, true]
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('object', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: {
            a: 'foo',
            b: 42,
            c: true
          }
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('null', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: null
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('undefined', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: undefined
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('events', () => {
    const Test = {
      name: 'Test',
      render(h: Function): any {
        return h()
      }
    }

    it('ports input event from the field', () => {
      const wrapper = shallowMount(BaseInput, {
        propsData: {
          value: 'str',
          availableTypes: ['string']
        },
        stubs: {
          InputString: Test
        }
      })

      wrapper.find(Test).vm.$emit('input', 'updated')
      expect(wrapper.emitted('input')[0][0]).toBe('updated')
    })
  })

  describe('will emit input event with empty value when the type is changed', () => {
    function changeType(wrapper: Wrapper<any>, type: ComponentDataType): void {
      const select = wrapper.find(BaseSelect)
      select.vm.$emit('change', type)
    }

    let wrapper: Wrapper<any>
    beforeEach(() => {
      wrapper = shallowMount(BaseInput, {
        propsData: {
          value: 'str'
        }
      })
    })

    it('string', () => {
      changeType(wrapper, 'string')
      expect(wrapper.emitted('input')[0][0]).toBe('')
    })

    it('number', () => {
      changeType(wrapper, 'number')
      expect(wrapper.emitted('input')[0][0]).toBe(0)
    })

    it('boolean', () => {
      changeType(wrapper, 'boolean')
      expect(wrapper.emitted('input')[0][0]).toBe(false)
    })

    it('array', () => {
      changeType(wrapper, 'array')
      expect(wrapper.emitted('input')[0][0]).toEqual([])
    })

    it('object', () => {
      changeType(wrapper, 'object')
      expect(wrapper.emitted('input')[0][0]).toEqual({})
    })

    it('null', () => {
      changeType(wrapper, 'null')
      expect(wrapper.emitted('input')[0][0]).toEqual(null)
    })

    it('undefined', () => {
      changeType(wrapper, 'undefined')
      expect(wrapper.emitted('input')[0][0]).toEqual(undefined)
    })
  })
})
