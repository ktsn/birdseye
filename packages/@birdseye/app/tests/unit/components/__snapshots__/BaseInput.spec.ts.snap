// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseInput rendering removes input 1`] = `
<div class="base-input">
  <!---->
  <select class="select-type">
    <option value="string">string</option>
    <option value="number">number</option>
    <option value="boolean">boolean</option>
    <option value="array">array</option>
    <option value="object">object</option>
  </select>
</div>
`;

exports[`BaseInput rendering removes type selector 1`] = `
<div class="base-input">
  <inputstring-stub value="foo"></inputstring-stub>
  <!---->
</div>
`;

exports[`BaseInput supported types array 1`] = `
<div class="base-input">
  <inputarray-stub value="foo,42,true"></inputarray-stub>
  <select class="select-type">
    <option value="string">string</option>
    <option value="number">number</option>
    <option value="boolean">boolean</option>
    <option value="array">array</option>
    <option value="object">object</option>
  </select>
</div>
`;

exports[`BaseInput supported types boolean 1`] = `
<div class="base-input">
  <inputboolean-stub value="true"></inputboolean-stub>
  <select class="select-type">
    <option value="string">string</option>
    <option value="number">number</option>
    <option value="boolean">boolean</option>
    <option value="array">array</option>
    <option value="object">object</option>
  </select>
</div>
`;

exports[`BaseInput supported types number 1`] = `
<div class="base-input">
  <inputnumber-stub value="123"></inputnumber-stub>
  <select class="select-type">
    <option value="string">string</option>
    <option value="number">number</option>
    <option value="boolean">boolean</option>
    <option value="array">array</option>
    <option value="object">object</option>
  </select>
</div>
`;

exports[`BaseInput supported types object 1`] = `
<div class="base-input">
  <inputobject-stub value="[object Object]"></inputobject-stub>
  <select class="select-type">
    <option value="string">string</option>
    <option value="number">number</option>
    <option value="boolean">boolean</option>
    <option value="array">array</option>
    <option value="object">object</option>
  </select>
</div>
`;

exports[`BaseInput supported types string 1`] = `
<div class="base-input">
  <inputstring-stub value="string value"></inputstring-stub>
  <select class="select-type">
    <option value="string">string</option>
    <option value="number">number</option>
    <option value="boolean">boolean</option>
    <option value="array">array</option>
    <option value="object">object</option>
  </select>
</div>
`;
